### PR TITLE
Propose an API for a charstream.

### PR DESCRIPTION
**Pull request by n1tram1**
Once we have a charstream API, we can start writing a lexer API.

### CODE REVIEW

File: include/unimplemented.h [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#ifndef UNIMPLEMENTED_H
+#define UNIMPLEMENTED_H
+
+#include <err.h>
+#include <stdlib.h>
+
+#define UNIMPLEMENTED()  \
+	errx(EXIT_FAILURE, "function %s in %s:%d not implemented", __FUNCTION__, __FILE__, __LINE__);
```
Comment (CohenArthur) [inline include/unimplemented.h:8]:
```suggestion
	errx(EXIT_FAILURE, "function %s in %s:%d not implemented", __FUNCTION__, __FILE__, __LINE__)
```
This way we have to enforce the use of a tailing semicolon, which in return helps clang-format

**Comment by CohenArthur** [comment]
Is a charstream really the best approach? Since we are simplifying the language, is there no higher lexer building block that we could use? For example, using parser generators or splitting the input in tokens directly, without the necessity to peek forward

**Comment by n1tram1** [comment]
> Is a charstream really the best approach?
I don't know if it is the best but it sure is useful.

> without the necessity to peek forward
afaik even with simple grammars, it's always nice to be able to peek characters

> using parser generators
we're not writing a recursive descent parser by hand ;( ? that's the best part of writing a shell imo

**Comment by CohenArthur** [comment]
> > we're not writing a recursive descent parser by hand ;( ? that's the best part of writing a shell imo

I mean, we can if you want, but it's a hassle haha

File: include/charstream.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * @file charstream.h
+ * Transform input from a file descriptor into a stream of characters.
+ */
+
+struct charstream {
+	/* Empty for now, it will be filled as implementation is written. */
+};
+
+int charstream_init(struct charstream *cs, int fd);
```
Comment (Skallwar) [inline include/charstream.h:15]: Maybe use FILE instead of file descriptor directly

File: include/unimplemented.h (hunk 1 above)
Comment (CohenArthur) [reply include/unimplemented.h:8]: It would also be nice to have this macro in a `utils.h` file, and we could just almost always include it

**Comment by Skallwar** [comment]
> I mean, we can if you want, but it's a hassle haha

If we don't there is not much to do to make a shell without HEREdocuments

**Review by Skallwar — APPROVED** [review]

File: include/unimplemented.h (hunk 1 above)
Comment (CohenArthur) [reply include/unimplemented.h:8]: How do you feel about this @n1tram1 ?

**Review by CohenArthur — COMMENTED** [review]
Looks solid!

File: include/unimplemented.h (hunk 1 above)
Comment (n1tram1) [reply include/unimplemented.h:8]: sounds good to me

**Review by CohenArthur — APPROVED** [review]

**Pull request merged**